concepts/websockets.textile

```diff
@@ -0,0 +1,591 @@
+
+h3(#moving-forward---other-things-you-might-consider). Moving forward - other things you might consider
+
+Basic implementation of a WebSocket server is really just the first stage of the process. Here are just few of the things you'd want to think about when taking things to the next level:
+
+* What framing extensions will you support, such as per-message deflation?
+* What degree of client interoperability are you aiming for?
+* Are messages being received in the same order they were sent, and if not, how can you prevent this from putting your application into an invalid state?
+* Do you need guarantees on message delivery, and if so, what strategies can you implement to this end?
+* How many connections are active on your server?
+* Are any connections hogging all of the server's resources?
+* Are any connections idle and should ideally be dropped?
+* What is the duration of the average connection lifespan?
+* Are connections being dropped prematurely/unexpectedly, and if so, how can you retain diagnostic data to explain why?
+* Are you experiencing brief connection spikes ever, and if so, what is the performance impact on your server?
+* How much bandwidth is being used overall, and how is it impacting your budget?
+* Is your server's capacity near its limit, and if not, how soon will that threshold be reached?
+* How will you automatically add additional server capacity if and when it is needed?
+
+Think about the messaging protocols available, such as MQTT, WAMP, etc., and whether they can provide a solution to some of these questions. Consider existing libraries and frameworks, and the additional features they offer beyond simple, bare-bones management of WebSocket connections. If you have a particular need to scale, and limited manpower or expertise to do so effectively, consider leveraging "cloud-based realtime messaging solutions":#ably-realtime-messaging-as-a-service that have already solved these problems for you.
+
+h2(#websocket-libraries-you-can-use-right-now). WebSocket libraries you can use right now
+
+There are two primary classes of WebSocket libraries; those that implement the protocol and leave the rest to the developer, and those that build on top of the protocol with various additional features commonly required by realtime messaging applications, such as restoring lost connections, pub/sub and channels, authentication, authorization, etc. The latter variety often require that their own libraries be used on the client side, rather than just using the raw @WebSocket@ API provided by the browser. As such, it becomes important to make sure you're happy with how they work and what they're offering. You may find yourself locked into your chosen solution's way of doing things once it has been integrated into your architecture, and any issues with reliability, performance and extensibility may come back to bite you.
+
+I'll start with a list of those that fall into the first of the above two categories;
```

Comment:
Please mention (perhaps repeatedly) that these are open sourced libraries.

Reply:
I have added a note in italics, appearing before the first subheading.